### ISSUE TITLE
Torch Xla 1.12 does not work on TPU-vm v3-32

### ISSUE THREAD

**Issue by jianguoz** [pytorch/xla#4103]
## ❓ Questions and Help
Hi, I have tried several ways to install tpu-pytorch `1.12` on TPU-VM, either `gcr.io/tpu-pytorch/xla:r1.12_3.8_tpuvm` or the pre-built wheels does not work.  

Install commands through docker image:
`
sudo docker pull gcr.io/tpu-pytorch/xla:r1.12_3.8_tpuvm

sudo docker run --privileged -v "/lib/libtpu.so:/lib/libtpu.so" --shm-size 16G -it --entrypoint=/bin/bash

export XRT_TPU_CONFIG="localservice;0;localhost:51011"
`

Install commands through pip:
`
pip install torch==1.12 https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch_xla-1.12-cp38-cp38-linux_x86_64.whl

export XRT_TPU_CONFIG="localservice;0;localhost:51011"
`

Below are the error messages when I run:
either 
`
python3 -c 'import torch_xla'
`
or examples in [huggingface torch xla](https://github.com/huggingface/transformers/tree/main/examples/pytorch) or examples in [Google cloud examples](https://cloud.google.com/tpu/docs/pytorch-pods). 

2022-10-18 09:11:20.075933: F ./tensorflow/core/tpu/tpu_library_init_fns.inc:100] TpuEmbeddingEngine_CollateMemory not available in this library.
https://symbolize.stripped_domain/r/?trace=7fba61c0600b,7fba61c0608f,7fb9421df0ea,7fb95094c046,7fb95094c616,7fb9424a2022,7fba61fc8b99&map= 
*** SIGABRT received by PID 31731 (TID 31731) on cpu 47 from PID 31731; stack trace: ***
PC: @     0x7fba61c0600b  (unknown)  raise
    @     0x7fb981d422f3        992  (unknown)
    @     0x7fba61c06090  (unknown)  (unknown)
    @     0x7fb9421df0eb        448  tensorflow::tpu::(anonymous namespace)::SetTpuOpsStructFns()
    @     0x7fb95094c047        128  tensorflow::tpu::InitializeTpuLibrary()
    @     0x7fb95094c617        592  tensorflow::tpu::FindAndLoadTpuLibrary()
    @     0x7fb9424a2023         32  _GLOBAL__sub_I_tpu_api_dlsym_initializer.cc
    @     0x7fba61fc8b9a  (unknown)  (unknown)
https://symbolize.stripped_domain/r/?trace=7fba61c0600b,7fb981d422f2,7fba61c0608f,7fb9421df0ea,7fb95094c046,7fb95094c616,7fb9424a2022,7fba61fc8b99&map=5920735bb186a93a82e10840f91bc184:7fb96d54f000-7fb9820c5fb0 
E1018 09:11:20.504434   31731 coredump_hook.cc:366] RAW: Remote crash data gathering hook invoked.
E1018 09:11:20.504464   31731 coredump_hook.cc:412] RAW: Skipping coredump since rlimit was 0 at process start.
E1018 09:11:20.504485   31731 client.cc:234] RAW: Coroner client retries enabled (b/136286901), will retry for up to 30 sec.
E1018 09:11:20.504495   31731 coredump_hook.cc:473] RAW: Sending fingerprint to remote end.
E1018 09:11:20.504505   31731 coredump_socket.cc:118] RAW: Stat failed errno=2 on socket /var/google/services/logmanagerd/remote_coredump.socket
E1018 09:11:20.504520   31731 coredump_hook.cc:477] RAW: Cannot send fingerprint to Coroner: [NOT_FOUND] Missing crash reporting socket. Is the listener running?
E1018 09:11:20.504526   31731 coredump_hook.cc:551] RAW: Discarding core.
E1018 09:11:20.508294   31731 process_state.cc:774] RAW: Raising signal 6 with default behavior

I then  update libtpu-nightly to the latest version through either
`
pip install --upgrade libtpu-nightly -f https://storage.googleapis.com/jax-releases/libtpu_releases.html
` 
or
`
pip3 install torch_xla[tpuvm]
`

Below are the errors:
`2022-10-18 09:18:24.090347: F ./tensorflow/core/tpu/tpu_library_init_fns.inc:101] TpuEmbeddingEngine_ConfigureCommunication not available in this library.
Aborted (core dumped)`.

or 
E1018 11:09:24.942148   50578 coredump_hook.cc:366] RAW: Remote crash data gathering hook invoked.
E1018 11:09:24.942173   50578 coredump_hook.cc:412] RAW: Called via ReportEvent and disabled coredump
E1018 11:09:24.942184   50578 client.cc:234] RAW: Coroner client retries enabled (b/136286901), will retry for up to 30 sec.
E1018 11:09:24.942189   50578 coredump_hook.cc:473] RAW: Sending fingerprint to remote end.
E1018 11:09:24.942193   50578 coredump_socket.cc:118] RAW: Stat failed errno=2 on socket /var/google/services/logmanagerd/remote_coredump.socket
E1018 11:09:24.942202   50578 coredump_hook.cc:477] RAW: Cannot send fingerprint to Coroner: [NOT_FOUND] Missing crash reporting socket. Is the listener running?
E1018 11:09:24.942206   50578 coredump_hook.cc:551] RAW: Discarding core.
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
https://symbolize.stripped_domain/r/?trace=7f3ecde7000b,7f3ecde7008f&map= 
*** SIGABRT received by PID 49843 (TID 50162) on cpu 27 from PID 49843; stack trace: ***
PC: @     0x7f3ecde7000b  (unknown)  raise
    @     0x7f3dee7e32f3        992  (unknown)
    @     0x7f3ecde70090  (unknown)  (unknown)
https://symbolize.stripped_domain/r/?trace=7f3ecde7000b,7f3dee7e32f2,7f3ecde7008f&map=5920735bb186a93a82e10840f91bc184:7f3dd9ff0000-7f3deeb66fb0 
E1018 11:09:26.922979   50162 coredump_hook.cc:366] RAW: Remote crash data gathering hook invoked.
E1018 11:09:26.922993   50162 coredump_hook.cc:412] RAW: Skipping coredump since rlimit was 0 at process start.
E1018 11:09:26.923002   50162 client.cc:234] RAW: Coroner client retries enabled (b/136286901), will retry for up to 30 sec.
E1018 11:09:26.923010   50162 coredump_hook.cc:473] RAW: Sending fingerprint to remote end.
E1018 11:09:26.923018   50162 coredump_socket.cc:118] RAW: Stat failed errno=2 on socket /var/google/services/logmanagerd/remote_coredump.socket
E1018 11:09:26.923025   50162 coredump_hook.cc:477] RAW: Cannot send fingerprint to Coroner: [NOT_FOUND] Missing crash reporting socket. Is the listener running?
E1018 11:09:26.923033   50162 coredump_hook.cc:551] RAW: Discarding core.
E1018 11:09:27.311076   50162 process_state.cc:774] RAW: Raising signal 6 with default behavior
Aborted (core dumped)

Can you help me to check the errors? Thanks

**Comment by hjm-aws** [comment]
@jianguoz I didn't try 1.12 on TPU. Can we keep the conversation in #4102 only about #4102? If you can remove the comment there it would be helpful. Thanks!

**Comment by wonjoolee95** [comment]
@jianguoz, thanks for the detailed info. Could you share the `--version` type you used (`tpu-vm-pt-1.11`, etc) when you created the TPUVM? You can also share the entire `glcoud` command you used to created the TPUVM.

**Comment by jianguoz** [comment]
> @jianguoz I didn't try 1.12 on TPU. Can we keep the conversation in #4102 only about #4102? If you can remove the comment there it would be helpful. Thanks!

Removed. Thanks:)

**Comment by jianguoz** [comment]
@wonjoolee95 Thanks for your quick help! I am using `tpu-vm-pt-1.12` for FSDP.  Below is one of the simplest versions:

`
TPU_NAME=tpu-jz-32  # change to your TPU name
ZONE=us-east1-d # change to your TPU zone
ACCELERATOR_TYPE=v3-32  # you can also try out larger TPU pods
RUNTIME_VERSION=tpu-vm-pt-1.12  # the XLA FSDP interface is supported in PyTorch/XLA

gcloud alpha compute tpus tpu-vm create ${TPU_NAME} \
  --zone ${ZONE} \
  --accelerator-type ${ACCELERATOR_TYPE} \
  --version ${RUNTIME_VERSION}
--metadata startup-script='#! /bin/bash
cd /usr/share/
git clone --recursive https://github.com/pytorch/pytorch
cd pytorch/
git clone --recursive https://github.com/pytorch/xla.git
EOF'
`

There are also other metadata (to install packages like transformers) information not listed here.  

Testing commands following [Scaling PyTorch models on Cloud TPUs with FSDP](https://pytorch.org/blog/scaling-pytorch-models-on-cloud-tpus-with-fsdp/) or examples in [Google cloud examples](https://cloud.google.com/tpu/docs/pytorch-pods). E.g.,
on single node v3-8:
`
python3 /usr/share/pytorch/xla/test/test_train_mp_mnist_fsdp_with_ckpt.py \
  --batch_size 16 --drop_last --num_epochs 2 \
  --use_nested_fsdp
`
or on v3-32 
`
sudo python3 -m torch_xla.distributed.xla_dist \
 --tpu=tpu-jz-32 --restart-tpuvm-pod-server -- python3 /usr/share/pytorch/xla/test/test_train_mp_imagenet.py \
 --fake_data --model=resnet50 --num_epochs=1
`

Note that when I change the version to `tpu-vm-pt-1.10`, everything is okay with torch==1.10 and torch_xla 1.10. But this version does not support scaling with FSDP

**Comment by muellerzr** [comment]
Also getting this issue over here toying with things for Accelerate :) 

Full reproducible steps:
**Configuration**:
- v3-32
- Zone: europe-west4-a
- Software: tpu-vm-pt-1.12 
- Architecture: TPU VM

**On main machine**:
1. `sudo docker pull gcr.io/tpu-pytorch/xla:nightly_3.8_tpuvm`
2. `sudo docker run --privileged  --shm-size 16G --name tpuvm_docker -it -d  gcr.io/tpu-pytorch/xla:nightly_3.8_tpuvm`
3. `sudo docker exec --privileged -it tpuvm_docker /bin/bash`

**In VM**:
Output from checking `libtpu`, `LD_PRELOAD`, and `accel` in that order:
`libtpu`: 
```
libtpu
libtpu_nightly-0.1.dev20220930.dist-info
```
`LD_PRELOAD`:
```
/usr/lib/x86_64-linux-gnu/libtcmalloc.so.4
```
`accel`:
```
accel0
accel1
accel2
accel3
```
1. `export XRT_TPU_CONFIG="localservice;0;localhost:51011"`
2. `echo 'print("Hello world from the TPU!")' >> test.py`
3. `python -m torch_xla.distributed.xla_dist --tpu=zach-accelerate -- python3 test.py`

## Error:

```bash
Traceback (most recent call last):
  File "/root/anaconda3/envs/pytorch/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/root/anaconda3/envs/pytorch/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/root/anaconda3/envs/pytorch/lib/python3.8/site-packages/torch_xla/distributed/xla_dist.py", line 661, in <module>
    cluster = cluster_resolver.get_cluster()
  File "/root/anaconda3/envs/pytorch/lib/python3.8/site-packages/torch_xla/distributed/cluster.py", line 462, in get_cluster
    service_workers = self.get_tpu_workers(as_client_worker=False)
  File "/root/anaconda3/envs/pytorch/lib/python3.8/site-packages/torch_xla/distributed/cluster.py", line 445, in get_tpu_workers
    xu.parallel_work(len(self._tpus), add_tpu_worker, self._tpus)
  File "/root/anaconda3/envs/pytorch/lib/python3.8/site-packages/torch_xla/utils/utils.py", line 293, in parallel_work
    return [res for res in results]  # Iterating to re-raise any exceptions
  File "/root/anaconda3/envs/pytorch/lib/python3.8/site-packages/torch_xla/utils/utils.py", line 293, in <listcomp>
    return [res for res in results]  # Iterating to re-raise any exceptions
  File "/root/anaconda3/envs/pytorch/lib/python3.8/concurrent/futures/_base.py", line 611, in result_iterator
    yield fs.pop().result()
  File "/root/anaconda3/envs/pytorch/lib/python3.8/concurrent/futures/_base.py", line 439, in result
    return self.__get_result()
  File "/root/anaconda3/envs/pytorch/lib/python3.8/concurrent/futures/_base.py", line 388, in __get_result
    raise self._exception
  File "/root/anaconda3/envs/pytorch/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/root/anaconda3/envs/pytorch/lib/python3.8/site-packages/torch_xla/distributed/cluster.py", line 408, in add_tpu_worker
    if ctc.state() != 'READY':
  File "/root/anaconda3/envs/pytorch/lib/python3.8/site-packages/cloud_tpu_client/client.py", line 274, in state
    return self._get_tpu_property('state')
  File "/root/anaconda3/envs/pytorch/lib/python3.8/site-packages/cloud_tpu_client/client.py", line 245, in _get_tpu_property
    metadata = self._fetch_cloud_tpu_metadata()
  File "/root/anaconda3/envs/pytorch/lib/python3.8/site-packages/cloud_tpu_client/client.py", line 239, in _fetch_cloud_tpu_metadata
    raise ValueError("Could not lookup TPU metadata from name '%s'. Please "
ValueError: Could not lookup TPU metadata from name 'zach-accelerate'. Please doublecheck the tpu argument in the TPUClusterResolver constructor. Exception: <HttpError 403 when requesting https://tpu.googleapis.com/v1/projects/******/locations/europe-west4-a/nodes/zach-accelerate?alt=json returned "Request had insufficient authentication scopes.". Details: "[{'@type': 'type.googleapis.com/google.rpc.ErrorInfo', 'reason': 'ACCESS_TOKEN_SCOPE_INSUFFICIENT', 'domain': 'googleapis.com', 'metadata': {'method': 'google.cloud.tpu.v1.Tpu.GetNode', 'service': 'tpu.googleapis.com'}}]">
```

Given mine and @jianguoz's errors differ slightly I can open a different issue if that's desired :)

**Comment by wonjoolee95** [comment]
Hey @jianguoz, I was able to follow the below steps to get the fsdp training complete successfully on a v3-8:
```
# Create v3-8
gcloud compute tpus tpu-vm create wonjoo-debug-v3-8 \
--zone europe-west4-a \
--accelerator-type v3-8  \
--version tpu-vm-pt-1.12 \
--project=my-project

# SSH into the TPUVM
gcloud compute tpus tpu-vm ssh wonjoo-debug-v3-8 \
--zone=europe-west4-a \
--project=my-project

# Make PyTorch direcotry and clone PyTorch/XLA for testing (this can also be included in the start-up script in the TPU creation command)
mkdir -p ~/pytorch && cd ~/pytorch
git clone --recursive -b r1.12 https://github.com/pytorch/xla.git
cd ~/

# Set env variables for TPUVM and run the training
export XRT_TPU_CONFIG="localservice;0;localhost:51011"
python3 ~/pytorch/xla/test/test_train_mp_mnist_fsdp_with_ckpt.py \
  --batch_size 16 --drop_last --num_epochs 2 \
  --use_nested_fsdp
```
The output of the training should look something like:
```
wonjoo@t1v-n-95fcc207-w-0:~$ python3 ~/pytorch/xla/test/test_train_mp_mnist_fsdp_with_ckpt.py   --batch_size 16 --drop_last --num_epochs 2   --use_nested_fsdp
Epoch 1 train begin 21:54:04
...
Epoch 2 train end 21:54:35
Epoch 2 test end 21:54:41, Accuracy=98.96
checkpoint saved to /tmp/mnist-fsdp/final_ckpt_rank-00000003-of-00000008.pth
checkpoint saved to /tmp/mnist-fsdp/final_ckpt_rank-00000000-of-00000008.pth
checkpoint saved to /tmp/mnist-fsdp/final_ckpt_rank-00000001-of-00000008.pth
checkpoint saved to /tmp/mnist-fsdp/final_ckpt_rank-00000002-of-00000008.pth
checkpoint saved to /tmp/mnist-fsdp/final_ckpt_rank-00000004-of-00000008.pth
checkpoint saved to /tmp/mnist-fsdp/final_ckpt_rank-00000005-of-00000008.pth
checkpoint saved to /tmp/mnist-fsdp/final_ckpt_rank-00000006-of-00000008.pth
checkpoint saved to /tmp/mnist-fsdp/final_ckpt_rank-00000007-of-00000008.pth
found 8 checkpoint files in /tmp/mnist-fsdp/final_ckpt_rank-*-of-*.pth
saved consolidated model to /tmp/mnist-fsdp/final_ckpt_consolidated.pth
Checkpoint consolidated, Accuracy=98.98 (note: it can be slightly different from the final training accuracy due to non-sync BatchNorm2d in the model)
Max Accuracy: 98.96%
wonjoo@t1v-n-95fcc207-w-0:~$ 
```
Couple of things to note:
1. This image is for `1.12`, so we also need to checkout the `r1.12` branch when cloning PyTorch/XLA
2. Setting up env vars for `XRT_TPU_CONFIG`

---
For v3-32 (pod), most of the steps should look the same:
```
# Create v3-32
gcloud compute tpus tpu-vm create wonjoo-debug-v3-32 \
--zone europe-west4-a \
--accelerator-type v3-32 \
--version tpu-vm-pt-1.12 \
--project=my-project \
--metadata startup-script='#! /bin/bash
cd /usr/share/
git clone --recursive https://github.com/pytorch/pytorch
cd /usr/share/pytorch
git clone --recursive -b r1.12 https://github.com/pytorch/xla.git
EOF'

# Note that we have to wait here a little bit until the start-up script finishes. Usually takes a couple a minutes. 

# SSH into the TPUVM
gcloud compute tpus tpu-vm ssh wonjoo-debug-v3-32 \
--zone=europe-west4-a \
--project=my-project \
-- -o ProxyCommand='corp-ssh-helper %h %p'

# Configure for pod use (enter empty when asked for a passphrase)
gcloud compute config-ssh

# Run the training
python3 -m torch_xla.distributed.xla_dist \
  --tpu=${TPU_NAME} \
  --restart-tpuvm-pod-server \
  -- python3 /usr/share/pytorch/xla/test/test_train_mp_imagenet.py \
  --fake_data \
  --model=resnet50 \
  --num_epochs=1 
```
Coupe of things to note on the pod use case:
1. We need to wait until the start-up script completes for each worker. For our use case, we can ssh into one of the workers and check if `PyTorch/XLA` cloning has finished at `/usr/share/pytorch/xla`. 
2. We need to run the command `gcloud compute config-ssh` to configure for pod.

**Comment by jianguoz** [comment]
@wonjoolee95 Thanks for your detailed reply. I could run `torch_xla 1.12` on v3-32 when checkouting the r1.12 branch when cloning PyTorch/XLA. I found there are still two issues:

1. torch_xla 1.12 is not compatible with [huggingface torch_xla examples](https://github.com/huggingface/transformers/tree/main/examples/pytorch).  Below are the errors:
`2022-10-20 01:58:28.567994: F ./tensorflow/core/tpu/tpu_library_init_fns.inc:100] TpuEmbeddingEngine_CollateMemory not available in this library.`

2. It does not directly support  torch_xla commands such as `python3 -c 'import torch_xla.core.xla_model as xm; print(xm.get_xla_supported_devices())'` on v3-32.  We must need to use 
'
python3 -m torch_xla.distributed.xla_dist \
  --tpu=${TPU_NAME} \
  --restart-tpuvm-pod-server \
  -- `python3 -c 'import torch_xla.core.xla_model as xm; print(xm.get_xla_supported_devices())'
'
Otherwise it will raise errors.

Could you please check these issues? Thanks very much:)

**Comment by JackCaoG** [comment]
for 1
```
2022-10-20 01:58:28.567994: F ./tensorflow/core/tpu/tpu_library_init_fns.inc:100] TpuEmbeddingEngine_CollateMemory not available in this library.
```

it is an known issue, uninstall the `tf` on the tpu-vm and install the `tf-cpu` instead

for 2

can you post the error message?

**Comment by jianguoz** [comment]
@JackCaoG Super thanks for the solutions of Q1:). Regarding 2, below is the error message:
```
2022-10-20 08:25:07.885109: W tensorflow/core/distributed_runtime/rpc/grpc_remote_master.cc:157] RPC failed with status = "UNAVAILABLE: Socket closed" and grpc_error_string = "{"created":"@1666254307.884917170","description":"Error received from peer ipv4:127.0.0.1:51011","file":"external/com_github_grpc_grpc/src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"Socket closed","grpc_status":14}", maybe retrying the RPC
```

Packages installed in TPU-VM are:
```
Package                      Version             
---------------------------- --------------------
absl-py                      1.1.0               
accelerate                   0.13.2              
aiohttp                      3.8.3               
aiosignal                    1.2.0               
astunparse                   1.6.3               
async-timeout                4.0.2               
attrs                        19.3.0              
Automat                      0.8.0               
blinker                      1.4                 
cachetools                   5.2.0               
certifi                      2022.6.15           
chardet                      3.0.4               
charset-normalizer           2.1.0               
Click                        7.0                 
cloud-init                   22.2                
cloud-tpu-client             0.10                
colorama                     0.4.3               
command-not-found            0.3                 
configobj                    5.0.6               
constantly                   15.1.0              
contourpy                    1.0.5               
cryptography                 2.8                 
cycler                       0.11.0              
Cython                       0.29.14             
datasets                     2.6.1               
dbus-python                  1.2.16              
dill                         0.3.5.1             
distlib                      0.3.4               
distro                       1.4.0               
distro-info                  0.23ubuntu1         
docker-pycreds               0.4.0               
entrypoints                  0.3                 
evaluate                     0.3.0               
filelock                     3.7.1               
fire                         0.4.0               
flatbuffers                  2.0                 
fonttools                    4.37.4              
frozenlist                   1.3.1               
fsspec                       2022.10.0           
ftfy                         6.1.1               
future                       0.18.2              
gast                         0.4.0               
gitdb                        4.0.9               
GitPython                    3.1.29              
google-api-core              1.31.6              
google-api-python-client     1.8.0               
google-auth                  2.9.0               
google-auth-httplib2         0.1.0               
google-auth-oauthlib         0.4.6               
google-cloud-core            2.3.2               
google-cloud-storage         2.5.0               
google-crc32c                1.5.0               
google-pasta                 0.2.0               
google-resumable-media       2.4.0               
googleapis-common-protos     1.56.3              
grpcio                       1.47.0              
h5py                         3.7.0               
httplib2                     0.20.4              
huggingface-hub              0.10.1              
hyperlink                    19.0.0              
idna                         3.3                 
importlib-metadata           4.12.0              
incremental                  16.10.1             
intel-openmp                 2022.1.0            
Jinja2                       2.10.1              
joblib                       1.2.0               
jsonlines                    3.1.0               
jsonpatch                    1.22                
jsonpointer                  2.0                 
jsonschema                   3.2.0               
keras                        2.10.0              
Keras-Applications           1.0.8               
Keras-Preprocessing          1.1.2               
keyring                      18.0.1              
kiwisolver                   1.4.4               
language-selector            0.1                 
launchpadlib                 1.10.13             
lazr.restfulclient           0.14.2              
lazr.uri                     1.0.3               
libclang                     14.0.1              
libtpu-nightly               0.1.dev20220615     
Markdown                     3.3.7               
MarkupSafe                   1.1.0               
matplotlib                   3.6.1               
mkl                          2022.1.0            
mkl-include                  2022.1.0            
mock                         4.0.3               
more-itertools               4.2.0               
multidict                    6.0.2               
multiprocess                 0.70.13             
netifaces                    0.10.4              
nltk                         3.7                 
numpy                        1.23.0              
oauth2client                 4.1.3               
oauthlib                     3.1.0               
opt-einsum                   3.3.0               
packaging                    21.3                
pandas                       1.5.1               
pathtools                    0.1.2               
pexpect                      4.6.0               
Pillow                       9.1.1               
pip                          20.0.2              
platformdirs                 2.5.2               
promise                      2.3                 
protobuf                     3.19.6              
psutil                       5.9.3               
pyarrow                      9.0.0               
pyasn1                       0.4.8               
pyasn1-modules               0.2.8               
PyGObject                    3.36.0              
PyHamcrest                   1.9.0               
PyJWT                        1.7.1               
pymacaroons                  0.13.0              
PyNaCl                       1.3.0               
pyOpenSSL                    19.0.0              
pyparsing                    3.0.9               
pyrsistent                   0.15.5              
pyserial                     3.4                 
python-apt                   2.0.0+ubuntu0.20.4.7
python-dateutil              2.8.2               
python-debian                0.1.36ubuntu1       
python-dotenv                0.20.0              
pytz                         2022.1              
PyYAML                       5.4.1               
regex                        2022.9.13           
requests                     2.28.1              
requests-oauthlib            1.3.1               
requests-unixsocket          0.2.0               
responses                    0.18.0              
rsa                          4.8                 
scikit-learn                 1.1.2               
scipy                        1.9.3               
SecretStorage                2.3.1               
sentencepiece                0.1.97              
sentry-sdk                   1.9.10              
service-identity             18.1.0              
setproctitle                 1.3.2               
setuptools                   62.6.0              
shortuuid                    1.0.9               
simplejson                   3.16.0              
six                          1.16.0              
sklearn                      0.0                 
smmap                        5.0.0               
sos                          4.3                 
ssh-import-id                5.10                
systemd-python               234                 
tbb                          2021.6.0            
tensorboard                  2.10.1              
tensorboard-data-server      0.6.1               
tensorboard-plugin-wit       1.8.1               
tensorflow-estimator         2.10.0              
tensorflow-io-gcs-filesystem 0.26.0              
termcolor                    1.1.0               
tf-cpu                       0.0.0               
threadpoolctl                3.1.0               
timm                         0.6.11              
tokenizers                   0.13.1              
torch                        1.12.0              
torch-xla                    1.12                
torchvision                  0.13.0              
tqdm                         4.64.1              
transformers                 4.23.1              
Twisted                      18.9.0              
typing-extensions            4.2.0               
ubuntu-advantage-tools       27.8                
ufw                          0.36                
unattended-upgrades          0.1                 
uritemplate                  3.0.1               
urllib3                      1.26.9              
virtualenv                   20.15.1             
wadllib                      1.3.3               
wandb                        0.13.4              
wcwidth                      0.2.5               
Werkzeug                     2.1.2               
wheel                        0.34.2              
wrapt                        1.14.1              
xxhash                       3.1.0               
yarl                         1.8.1               
zipp                         1.0.0               
zope.interface               4.7.1    
```
